### PR TITLE
Validate fields against datatype 

### PR DESCRIPTION
**Pull request by gidsg**
* Adds validation of fields against their datatype when submitting an update
* Validates that the key is populated
screen shot 2017-09-18 at 15 44 00

### CODE REVIEW

File: app/views/register/shared/_form-error-summary.haml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+%div.error-summary{:role => 'alert', :'aria-labelledby' => 'error-summary-heading', :tabindex => '-1'}
```
Comment (danmitchell-) [inline app/views/register/shared/_form-error-summary.haml]: I think we should move this outside of the "registers" directory. I would suggest the layouts directory

File: app/views/register/new.html.haml [hunk 2]
```diff
@@ -1,6 +1,9 @@
 = content_for :page_title, "New #{params[:register]}"
 = link_to 'Back', "/#{params[:register]}", class: 'link-back'
 
+- if flash.present?
```
Comment (danmitchell-) [inline app/views/register/new.html.haml]: We should shorten this to `= render partial: 'register/shared/form-error-summary' if flash.present?` but I think this should be in the application layout file instead

File: app/views/register/shared/_form.html.haml [hunk 3]
```diff
@@ -1,10 +1,14 @@
 = form_for :create, url: "/#{params[:register]}" do |f|
   - @register.fields.each do |field|
-    .form-group
-      = label_tag field, prepare_register_name(field), class: 'form-label form-label-bold'
+  - has_error = flash[field].present?
+    .form-group{:class => ('form-group-error' if has_error) }
+      %label{class: 'form-label form-label-bold'}
```
Comment (danmitchell-) [inline app/views/register/shared/_form.html.haml]: `%label.form-label.form-label-bold`

File: app/views/register/shared/_form.html.haml [hunk 4]
```diff
@@ -1,10 +1,14 @@
...
+        =prepare_register_name(field)
+        - if has_error
+          %span{class: 'error-message'} #{flash[field]}
       %span.form-hint #{get_description_for_register_field(field)}
       - if params[:action] == 'edit' && field == params[:register]
         = text_field_tag field, field != params[:register] ? @form[field] : @form['key'], class: 'form-control disabled'
       - else
-        = text_field_tag field, @form[field], class: 'form-control'
+        = text_field_tag field, @form[field], :class => [('form-control'), ('form-control-error' if has_error)]
```
Comment (danmitchell-) [inline app/views/register/shared/_form.html.haml]: This might be neater `= text_field_tag field, @form[field], class: "form-control #{'form-control-error' if has_error}`

File: app/controllers/register_controller.rb [hunk 5]
```diff
@@ -62,4 +72,11 @@ def create
     redirect_to action: 'index', register: params[:register]
   end
 
+  private
+
+  def initialize_controller
```
Comment (danmitchell-) [inline app/controllers/register_controller.rb]: hmmm I not sure this is rails convention but let me find out for sure

File: app/controllers/change_controller.rb [hunk 6]
```diff
@@ -74,7 +74,7 @@ def update
   end
 
   def post_to_register(register_name, rsf_body)
-    protocol = Rails.configuration.register_ssl ? protocol = 'https' : protocol = 'http'
+    protocol = Rails.configuration.register_ssl ? 'https' : 'http'
```
Comment (karlbaker02) [inline app/controllers/change_controller.rb]: This has been tested?

File: app/helpers/validation_helper.rb [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+          'string' => ValidationHelper::StringDatatype.new,
+          'point' => ValidationHelper::PointDatatype.new,
+          'url' => ValidationHelper::UrlDatatype.new,
+          'curie' => ValidationHelper::CurieDatatype.new,
+          'datetime' => ValidationHelper::DateDatatype.new
+      }.freeze
+    end
+
+    def get_form_errors(params, field_definitions)
+
```
Comment (karlbaker02) [inline app/helpers/validation_helper.rb]: Small point: could we get rid of this whitespace?

File: app/controllers/change_controller.rb (hunk 6 above)
Comment (gidsg) [reply app/controllers/change_controller.rb]: I have tested the syntax in IRB. This was a suggested refactor in RubyMine as protocol is needlessly redeclared.

Comment (karlbaker02) [reply app/controllers/change_controller.rb]: Makes total sense, just checking with my limited experience of Ruby.

File: app/controllers/register_controller.rb [hunk 8]
```diff
@@ -27,21 +28,30 @@ def edit
...
-                        :beta)
-
-    if @current_register_record != nil
-      @current_register_record = convert_register_json(@current_register_record)
+    errors = @data_validator.get_form_errors(params,@registers_client.get_register(params[:register], 'beta').get_field_definitions)
+    if errors.present?
+      errors.each { |k,v| flash[k] = v[:message] }
+      @register = get_register(params[:register])
+      @form = JSON.parse(params.to_json)
+      render :new
```
Comment (karlbaker02) [inline app/controllers/register_controller.rb]: Should we also return `false` after this line, like we do on line 50?

File: app/views/register/shared/_form.html.haml (hunk 4 above)
Comment (gidsg) [reply app/views/register/shared/_form.html.haml]: the syntax you suggested didn't work for me, not sure it is valid to have a conditional like that in the template.

File: app/controllers/register_controller.rb (hunk 8 above)
Comment (gidsg) [reply app/controllers/register_controller.rb]: Doesn't seem to make a functional difference either way. Not sure what we'd be indicating by doing this?

Comment (gidsg) [reply app/controllers/register_controller.rb]: @danmitchell- says the other style is wrong so I will remove that `return false`

File: app/controllers/change_controller.rb (hunk 6 above)
Comment (simonwo) [reply app/controllers/change_controller.rb]: This is legit. +1

**Comment by laurentqro** [comment]
Rails support model attributes validation out of the box: http://api.rubyonrails.org/classes/ActiveModel/Validations.html.

example:

```
class Person
  include ActiveModel::Validations

  attr_accessor :first_name, :last_name

  validates_each :first_name, :last_name do |record, attr, value|
    record.errors.add attr, 'starts with z.' if value.to_s[0] == ?z
  end
end
```

**Comment by karlbaker02** [comment]
@laurentqro This is great, but if I understand correctly the problem will be that the fields that need validating will be dynamic based upon which register it is, which means that we can't create an ActiveModel to represent this as it will not cater to all registers.

**Comment by simonwo** [comment]
Could you metaprogram yourself a model at runtime? E.g.

    register = @registers_client.get_register(params[:register], 'beta')
    register_model_class = Class.new do
      register.get_field_definitions.each do |field|
        attr_accessor field[:item]['field'].to_sym

        case field[:item]['datatype']
        when 'integer'
          validates field[:item]['field'].to_sym, numericality: { only_integer: true }
        when 'curie'
          validates field[:item]['field'].to_sym, format: /^\w+:\w+$/
        ...
      end
    end

Not sure if you think that's more or less horrible, but it lets you use the ActiveModel stuff.

File: app/controllers/register_controller.rb [hunk 9]
```diff
@@ -33,21 +34,30 @@ def edit
...
-    return true if params[:data_confirmed]
-    @register = get_register(params[:register])
-
-    @current_register_record = OpenRegister.record(params[:register].downcase,
-                        params[params[:register].downcase.to_sym],
-                        :beta)
-
-    if @current_register_record != nil
-      @current_register_record = convert_register_json(@current_register_record)
+    errors = @data_validator.get_form_errors(params,@registers_client.get_register(params[:register], 'beta').get_field_definitions)
```
Comment (govuklaurence) [inline app/controllers/register_controller.rb]: * Needs spaces after `,`
* It may be better to build a model with validations (ActiveModel) then ask it if it's valid -- that way the errors that come back will tell you exactly what's wrong with no extra code in `lib`
* Long line
* Might be a good idea to break up and use some intermediate vars

File: app/controllers/change_controller.rb (hunk 6 above)
Comment (govuklaurence) [reply app/controllers/change_controller.rb]: This works fine

File: app/helpers/validation_helper.rb [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        validation_success(value)
+      rescue ArgumentError
+        validation_failure("#{value} is not a valid integer")
+      end
+    end
+  end
+
+  class StringDatatype < Datatype
+    def validate(value)
+      validation_success(value)
```
Comment (govuklaurence) [inline app/helpers/validation_helper.rb]: No checking at all?

File: app/helpers/validation_helper.rb [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    end
+  end
+
+  class PointDatatype < Datatype
```
Comment (govuklaurence) [inline app/helpers/validation_helper.rb]: Duplicated methods

File: app/helpers/validation_helper.rb [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    def validate(value)
+      begin
+        uri = URI.parse(value)
+
+        if ((uri.kind_of?(URI::HTTP) || uri.kind_of?(URI::HTTPS)) && !uri.host.nil?)
+          validation_success(url)
+        else
+          validation_failure("#{url} is not a valid URL")
+        end
+      rescue
```
Comment (govuklaurence) [inline app/helpers/validation_helper.rb]: Better to be more specific with what errors are caught here

File: app/helpers/validation_helper.rb [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        YEAR_MONTH_DAY_HOURS_MINUTES_SECONDS,
+        YEAR_MONTH_DAY,
+        YEAR_MONTH,
+        YEAR
+    ].freeze
+
+    def validate(value)
+      parsed_date = nil
+
+      DATE_FORMATS.each { |date_format|
```
Comment (govuklaurence) [inline app/helpers/validation_helper.rb]: on multiline blocks, use `do ... end` instead of `{ .. }`

File: app/helpers/validation_helper.rb [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        begin
+          parsed_date = Date.strptime(value, date_format)
+          return validation_success(parsed_date)
```
Comment (govuklaurence) [inline app/helpers/validation_helper.rb]: Useless use of `return`

File: app/helpers/validation_helper.rb [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+          result[field_name] = { success: false, message: "Field #{field_name} is required" }
+        end
+      }
+      result
+    end
+  end
+
+  class Datatype
+    def validation_success(value)
+      { success: true, value: value}
```
Comment (govuklaurence) [inline app/helpers/validation_helper.rb]: The response schema of the validator changes based on success or failure

File: app/helpers/validation_helper.rb [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+          'point' => ValidationHelper::PointDatatype.new,
+          'url' => ValidationHelper::UrlDatatype.new,
+          'curie' => ValidationHelper::CurieDatatype.new,
+          'datetime' => ValidationHelper::DateDatatype.new
+      }.freeze
+    end
+
+    def get_form_errors(params, field_definitions)
+      result = {}
+      field_definitions.each{ |field|
```
Comment (govuklaurence) [inline app/helpers/validation_helper.rb]: `do ... end` on multiline blocks

File: app/controllers/register_controller.rb [hunk 17]
```diff
@@ -2,6 +2,7 @@ class RegisterController < ApplicationController
 
   include ApplicationHelper, RegisterHelper
 
+  before_action :initialize_controller
```
Comment (govuklaurence) [inline app/controllers/register_controller.rb]: Do you need the instance variables in every action?

**Comment by laurentqro** [comment]
@karlbaker02 what are the different fields? And what are the different kinds of registers?

**Comment by govuklaurence** [comment]
Metaprogramming shouldn’t be required, just allow attributes not to be present, but if they are, validate them

**Comment by gidsg** [comment]
@laurentqro You can see the datatypes we support here: https://datatype.alpha.openregister.org/records registers can use any arbitrary combinations of fields with any of these datatypes.

**Comment by simonwo** [comment]
@govuklaurence Does that require listing all of the possible attributes up front? A Register could have any arbitrary set of fields and this schema isn't known at code-writing time.

**Comment by gidsg** [comment]
OK, I have updated this as suggested now to use active record validation. Please take a look @karlbaker02 @simonwo

File: app/helpers/validation_helper.rb [hunk 18]
```diff
@@ -0,0 +1,123 @@
+module ValidationHelper
```
[the remark this answers is not in the document]
Comment (gidsg) [reply app/helpers/validation_helper.rb]: this is done now. The custom validators are in their own classes.

File: app/views/layouts/_form-error-summary.haml [hunk 19]
```diff
@@ -4,4 +4,4 @@
   %ul.error-summary-list
     - flash.each do |field, message|
       %li
-      = link_to message, "##{field}"
\ No newline at end of file
+      = link_to message, "##{field.dasherize}"
```
Comment (gidsg) [inline app/views/layouts/_form-error-summary.haml]: All this wrangling between `dasherize` and `underscore` is to convert from register fields with dashes e.g. `end-date` to something safe for a ruby symbol `end_date`

File: app/controllers/register_controller.rb [hunk 20]
```diff
@@ -33,19 +34,27 @@ def edit
...
+      @form = JSON.parse(params.to_json)
+      render :new
+    else
+      return true if params[:data_confirmed]
+      @register = get_register(params[:register])
+      @current_register_record = OpenRegister.record(register_name,
+                          params[register_name.to_sym],
+                          :beta)
+
+      if @current_register_record != nil
```
Comment (simonwo) [inline app/controllers/register_controller.rb]: What happens if it is `nil`? Will rendering `:confirm` explode with missing fields?

Comment (danmitchell-) [reply app/controllers/register_controller.rb]: This needs changing to just `if @current_register_record`

Comment (danmitchell-) [reply app/controllers/register_controller.rb]: actually this isn't doing anything now 🤔

File: app/helpers/validation_helper.rb [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+          is_key = field_sym == register_sym
+
+          case field[:item]['datatype']
+            # Custom validators are defined in app/validators
+          when 'integer'
+            validates field_sym, numericality: { only_integer: true, message: "%{value} is not an integer" }, allow_blank: true
+          when  'curie'
+            validates field_sym, format:{ with: /\A\w+:\w+\z/, message: "%{value} is not a valid curie"}, allow_blank: true
+          when 'string'
+            validates field_sym, presence: { message: "Field %{attribute} is required"}, allow_blank: !is_key
```
Comment (gidsg) [inline app/helpers/validation_helper.rb]: I am assuming the key will always be a string, so I only check strings to see if they should be mandatory fields. Let me know if this is not a correct assumption.

File: app/controllers/register_controller.rb [hunk 22]
```diff
@@ -33,19 +34,27 @@ def edit
...
+      @register = get_register(params[:register])
+      @current_register_record = OpenRegister.record(register_name,
+                          params[register_name.to_sym],
+                          :beta)
+
+      if @current_register_record
+        @current_register_record = convert_register_json(@current_register_record)
+      end
+
+      render 'confirm'
```
Comment (simonwo) [inline app/controllers/register_controller.rb]: Style: should this be a `Symbol`?

File: config/environments/production.rb [hunk 23]
```diff
@@ -3,7 +3,7 @@
 
   # Code is not reloaded between requests.
   config.cache_classes = true
-
+  config.autoload_paths += %W["#{config.root}/app/validators/"]
```
Comment (danmitchell-) [inline config/environments/production.rb]: We can move this into application.rb to save including in ever environment file

Comment (danmitchell-) [reply config/environments/production.rb]: every*
